python/lsst/meas/algorithms/setPrimaryFlags.py

```diff
@@ -0,0 +1,308 @@
+                          "and is not \"detected\" in a pseudo-filter (see config.pseudoFilterList)")
+            self.isPatchInnerKey = self.schema.addField(
+                "detect_isPatchInner", type="Flag",
+                doc="True if source is in the inner region of a coadd patch",
+            )
+            self.isTractInnerKey = self.schema.addField(
+                "detect_isTractInner", type="Flag",
+                doc="True if source is in the inner region of a coadd tract",
+            )
+        else:
+            primaryDoc = "True if source has no children and is not a sky source."
+        self.isPrimaryKey = self.schema.addField(
+            "detect_isPrimary", type="Flag",
+            doc=primaryDoc,
+        )
+
+        if "deblend_nChild" in schema.getNames():
+            self.includeDeblend = True
+            self.isDeblendedSourceKey = self.schema.addField(
+                "detect_isDeblendedSource", type="Flag",
+                doc=primaryDoc + " and is either an unblended isolated source or a "
+                                 "deblended child from a parent with 'deblend_nChild' > 1")
+            self.fromBlendKey = self.schema.addField(
+                "detect_fromBlend", type="Flag",
+                doc="This source is deblended from a parent with more than one child."
+            )
+            self.isIsolatedKey = self.schema.addField(
+                "detect_isIsolated", type="Flag",
+                doc="This source is not a part of a blend."
+            )
+            if "deblend_scarletFlux" in schema.getNames():
+                self.isDeblendedModelKey = self.schema.addField(
+                    "detect_isDeblendedModelSource", type="Flag",
+                    doc=primaryDoc + " and is a deblended child")
+            else:
+                self.isDeblendedModelKey = None
+
+    def run(self, sources, skyMap=None, tractInfo=None, patchInfo=None):
+        """Set isPrimary and related flags on sources.
+
+        For coadded imaging, the `isPrimary` flag returns True when an object
+        has no children, is in the inner region of a coadd patch, is in the
+        inner region of a coadd trach, and is not detected in a pseudo-filter
+        (e.g., a sky_object).
+        For single frame imaging, the isPrimary flag returns True when a
+        source has no children and is not a sky source.
+
+        Parameters
+        ----------
+        sources : `lsst.afw.table.SourceCatalog`
+            A sourceTable. Reads in centroid fields and an nChild field.
+            Writes is-patch-inner, is-tract-inner, and is-primary flags.
+        skyMap : `lsst.skymap.BaseSkyMap`
+            Sky tessellation object
+        tractInfo : `lsst.skymap.TractInfo`, optional
+            Tract object; required if ``self.isSingleFrame`` is False.
+        patchInfo : `lsst.skymap.PatchInfo`
+            Patch object; required if ``self.isSingleFrame`` is False.
```

Comment:
No need to specify "optional" in the docstring, since the function signature makes it clear that they are optional parameters.